scipy/spatial/transform/_rigidtransformation.pyx

```diff
@@ -0,0 +1,1899 @@
+               [-0.2381705, -0.87102639, -0.42963687, 0.19637636],
+               [0., 0., 0., 1.]])
+        >>> t.single
+        True
+        """
+        expcoords = np.asarray(expcoords, dtype=float)
+
+        if (expcoords.ndim not in [1, 2] or expcoords.shape[0] == 0
+                or expcoords.shape[-1] != 6):
+            raise ValueError(
+                "Expected `expcoords` to have shape (6,), or (N, 6), "
+                f"got {expcoords.shape}.")
+
+        single = expcoords.ndim == 1
+        expcoords = np.atleast_2d(expcoords)
+
+        rotations = Rotation.from_rotvec(expcoords[:, :3])
+
+        theta = np.linalg.norm(expcoords[:, :3], axis=-1)
+        ind_only_translation = theta == 0.0
+
+        translations = np.empty((len(expcoords), 3), dtype=float)
+
+        if not np.all(ind_only_translation):
+            theta[ind_only_translation] = 1.0
+            screw_axes = expcoords / theta[:, np.newaxis]
+
+            tms = theta - np.sin(theta)
+            cm1 = np.cos(theta) - 1.0
+            o0 = screw_axes[:, 0]
+            o1 = screw_axes[:, 1]
+            o2 = screw_axes[:, 2]
+            v0 = screw_axes[:, 3]
+            v1 = screw_axes[:, 4]
+            v2 = screw_axes[:, 5]
+            o01tms = o0 * o1 * tms
+            o12tms = o1 * o2 * tms
+            o02tms = o0 * o2 * tms
+            o0cm1 = o0 * cm1
+            o1cm1 = o1 * cm1
+            o2cm1 = o2 * cm1
+            o00tms = o0 * o0 * tms
+            o11tms = o1 * o1 * tms
+            o22tms = o2 * o2 * tms
+            translations[..., 0] = (-v0 * (o11tms + o22tms - theta)
+                + v1 * (o01tms + o2cm1)
+                + v2 * (o02tms - o1cm1))
+            translations[..., 1] = (v0 * (o01tms - o2cm1)
+                - v1 * (o00tms + o22tms - theta)
+                + v2 * (o0cm1 + o12tms))
+            translations[..., 2] = (v0 * (o02tms + o1cm1)
+                - v1 * (o0cm1 - o12tms)
+                - v2 * (o00tms + o11tms - theta))
+
+        translations[ind_only_translation] = expcoords[ind_only_translation, 3:]
+
+        # TODO alternatively, we could
+        # return cls.from_translation(translations) * cls.from_rotation(rotations)
+
+        matrix = np.empty((len(expcoords), 4, 4), dtype=float)
+        matrix[:, :3, :3] = rotations.as_matrix()
+        matrix[:, :3, 3] = translations
+        matrix[:, 3, :3] = 0.0
+        matrix[:, 3, 3] = 1.0
+        if single:
+            matrix = matrix[0]
+
+        return cls(matrix, normalize=False, copy=False)
+
+    @cython.embedsignature(True)
+    @classmethod
+    def from_dualquat(cls, dualquat, *, scalar_first=False):
+        """Initialize from a unit dual quaternion.
+
+        Unit dual quaternions encode orientation in a real unit quaternion
+        and translation in a dual quaternion. There is a double cover, i.e.,
+        the unit dual quaternions q and -q represent the same transformation.
```

Comment:
Dual quaternions are tricky, their unity property is not what people might think. 

Intuitively we want to normalize dual quaternions before converting them into the transformation matrix similarly to rotation quaternions.

Is there a known way to do it?

Reply:
The cool thing about dual quaternions is that their norm only depends on the norm of the real quaternion, since $\epsilon^2 = 0$ and the dual and real quaternion are orthogonal. So you only compute the norm of the first 4 elements and divide the whole 8d array by it.

Reply:
> and the dual and real quaternion are orthogonal

Exactly this, but arbitrary 8 numbers won't satisfy it. So we need normalize the real part and orthogonalize real and dual parts.

In principle, am I correct that it's better to do that in this function rather than assume that the given numbers indeed form a dual quaternion? I think the latter approach is bad for a computation package so we need to implement normalization.

Reply:
Ah OK, so that might be a bit more complicated.

There seems to be a solution: https://stackoverflow.com/questions/23174899/properly-normalizing-a-dual-quaternion

However, I am not exactly sure about this. What about just raising an error?

Reply:
Definitely it is less intuition here what this normalization means. But in order to raise an error we need to introduce a somewhat arbitrary threshold.

I have checked what is the consequences of lack of orthogonality. An obvious thing is that we get nonzero value for the scalar part of the translation (after the quaternion multiplication), this we can simply drop as currently implemented. However for the vector part (actual translation which is passed  to the constructor) the effect is absolutely non obvious, we simply get some numbers which I don't know how to intepret.

I'm inclined to do the normalization because it is better defined that validity checking. It can be formulated as "only orthogonal part of the dual quaternion is kept".

Reply:
I have an experimental implementation of the normalization procedure here: https://github.com/dfki-ric/pytransform3d/blob/1afe3570f5653e523e33d5c740e328a00d047bfa/pytransform3d/transformations/_dual_quaternion_operations.py#L9

We would need to implement the dual quaternion product for this, which we currently only use for this function. Other operations are trivial or exist already: component-wise quaternion conjugate and quaternion multiplication. It's definitely doable, but might also add some computational overhead. Do we want to do this every time?

Reply:
PR: https://github.com/scottshambaugh/scipy/pull/6

Reply:
Merged